sdk/eventhub/Azure.Messaging.EventHubs/samples/Sample11_MockingClientTypes.md

```diff
@@ -0,0 +1,382 @@
+
+// Create a mock of a partition event
+PartitionEvent samplePartitionEvent = new PartitionEvent(partitionContext, eventData);
+List<PartitionEvent> partitionEventList = new List<PartitionEvent>{ samplePartitionEvent };
+
+// Define a simple local method that returns an IAsyncEnumerable to use as the return for
+// ReadEventsAsync below
+async IAsyncEnumerable<PartitionEvent> mockReturn(PartitionEvent samplePartitionEvent)
+{
+    await Task.Yield();
+    yield return samplePartitionEvent;
+}
+
+// Use this PartitionEvent to mock a return from the consumer, because ReadEvents returns an IAsyncEnumerable a separate
+// method is needed to properly set up this method
+mockConsumer
+    .Setup(c => c.ReadEventsAsync(It.IsAny<CancellationToken>()))
+    .Returns(mockReturn(samplePartitionEvent));
+
+EventHubConsumerClient consumer = mockConsumer.Object;
+
+using CancellationTokenSource cancellationTokenSource = new CancellationTokenSource();
+
+await foreach (PartitionEvent receivedEvent in consumer.ReadEventsAsync(cancellationTokenSource.Token))
+{
+    Assert.AreEqual(receivedEvent.Data, eventData);
+}
+```
+
+## Receiving Batches using the `PartitionReceiver`
+
+The sample below illustrates how to mock a `PartitionReceiver`, and set up its `ReceiveBatchAsync` method to return a simple data batch. 
+
+Given the purpose of the `PartitionReceiver`, it is anticipated that most applications will have complex code surrounding their receivers, meaning mocking these data types can be especially helpful.
+
+```C# Snippet:EventHubs_Sample11_PartitionReceiverMock
+// Create a mock of the PartitionReceiver
+Mock<PartitionReceiver> mockReceiver = new Mock<PartitionReceiver>();
+
+// Create a list of events to receive from the PartitionReceiver
+List<EventData> receivedEvents = new List<EventData>();
+
+for (int i=0; i<10; i++)
+{
+    // Mocking an EventData instance, different arguments can simulate different potential
+    // outputs from the broker
+    EventData eventData = EventHubsModelFactory.EventData(
+        eventBody: new BinaryData($"Sample-Event-{i}"),
+        systemProperties: default,
+        partitionKey: "0",
+        sequenceNumber: default,
+        offset: default,
+        enqueuedTime: default);
+
+    receivedEvents.Add(eventData);
+}
+
+// Setup the mock to receive the mocked data batch when ReceiveBatchAsync is called
+// on a given partition
+mockReceiver
+    .Setup(r => r.ReceiveBatchAsync(
+        It.IsAny<int>(),
+        It.IsAny<TimeSpan>(),
+        It.IsAny<CancellationToken>()))
+    .ReturnsAsync(receivedEvents);
+
+PartitionReceiver receiver = mockReceiver.Object;
+
+using CancellationTokenSource cancellationTokenSource = new CancellationTokenSource();
+
+IEnumerable<EventData> receivedBatch = await receiver.ReceiveBatchAsync(
+    maximumEventCount: 10,
+    TimeSpan.FromSeconds(1),
+    cancellationTokenSource.Token);
+```
+
+## Mocking a custom event processor using `EventProcessor<Partition>` 
```

Comment:
This will be an interesting one that we may (or may not) want to include; if developers are rolling their own processor, I think we can assume that they're proficient with testing.   This sample gives them the tools for figuring out how to mock events and other internalized data types that they'd need to test.

That said, I think we should include one of these for the Event Processor Client that:

1) Tells people that they should only test their own handler code by calling it directly
2) Shows off the model factory + Event Data + calling a handler
3) Links back to this sample.